### DISCUSSION TITLE
v0.11.6 Testing checklist

### DISCUSSION THREAD

**Discussion by jorgeepc** [kubeshop/tracetest#2441] [Release testing]
# Testing Tracetest

This is the entire checklist on what we should do to assert that Tracetest is working fine on each version release. On each version release, we can copy the contents of this template checklist and open a Github Discussion to start each test.

On Tracetest, we work in two ways to test the entire system and guarantee that everything is working fine. One is through automatic tests, where we run unit, integration and end-to-end tests on Tracetest CLI, Web UI and Server.

Another source of tests is the manual tests that we execute on each release, simulating users, and checking if everything is ok on these tests.

## Manual Tests

On our manual tests, we aim to do some [sanity checks](https://en.wikipedia.org/wiki/Sanity_check) to see if the main features are working as expected. Usually, we run it on each Tracetest release.

### Tests to validate on Release Candidate

- Test server installation via CLI
  - [x] Docker Compose and no demo API
  - [x] Docker Compose and demo API
  - [x] Kubernetes and no demo API
  - [ ] Kubernetes and demo API
  - [x] Verify that Tracetest is not sending analytics if you have the `TRACETEST_DEV` env for docker install
  - [x] Verify that Tracetest is not sending analytics if you have the `TRACETEST_DEV` env for k8s install

- Test DataStore provisioning and basic test
  - [x] Open WebUI and go to `/settings` page. The provisioned Data Store should be selected
  - [x] Create a test on WebUI that calls a demo API (like [Pokeshop](https://docs.tracetest.io/live-examples/pokeshop/overview) or [Open Telemetry Store](https://docs.tracetest.io/live-examples/opentelemetry-store/overview)). This test should fetch traces correctly and run without errors

- Test Tracetest examples
  - [x] [Jaeger example](https://github.com/kubeshop/tracetest/tree/main/examples/tracetest-jaeger)
  - [x] [Tempo example](https://github.com/kubeshop/tracetest/tree/main/examples/tracetest-tempo)
  - [x] [New Relic example](https://github.com/kubeshop/tracetest/tree/main/examples/tracetest-new-relic)
  - [x] [Lightstep example](https://github.com/kubeshop/tracetest/tree/main/examples/tracetest-lightstep)
  - [x] [OpenSearch example](https://github.com/kubeshop/tracetest/tree/main/examples/tracetest-opensearch)
  - [x] [Elastic APM example](https://github.com/kubeshop/tracetest/tree/main/examples/tracetest-elasticapm-with-elastic-agent)
  - [x] [Datadog example](https://github.com/kubeshop/tracetest/tree/main/examples/tracetest-datadog)
  - [x] [Amazon X-Ray example](https://github.com/kubeshop/tracetest/tree/main/examples/tracetest-amazon-x-ray)

- Test specific features added/changed on this release:
  - [x] Validate new CLI `datastore` resource: `apply datastore`, `get datastore`, `delete datastore`
  - [x] Validate CLI `config` resource: `apply config`, `get config`
  - [x] Validate CLI `pollingprofile` resource: `apply pollingprofile`, `get pollingprofile`
  - [x] Validate CLI `demo` resource: `apply demo`, `get demo`, `delete demo`, `list demo`
  - [x] Validate webUI datastore CRUD
  - [x] Test HTTP OTLP server
  - [x] Create a test with an URL without scheme, it should default to http
  
### Tests to validate on Final Release

- Pipeline
  - [x] Check if our release pipeline on [Release Tracetest](https://github.com/kubeshop/tracetest/actions/workflows/release-version.yml) workflow on Github Actions worked correctly

- Test CLI updates
  - [x] MacOS via homebrew
  - [x] MacOS via curl script
  - [x] Windows via chocolatey
  - [x] Windows via manual download

- Docs
  - [x] Double check [Detailed installation](https://docs.tracetest.io/getting-started/detailed-installation) docs and see if everything is documented correctly
  
## Automatic Tests

Today Tracetest has 3 main components: a WebUI, a CLI and a Server.

### Web UI

- **Unit tests**: Run by executing `npm test` on `./web` folder
- **End-to-end tests**: Run using [cypress](https://www.cypress.io/) against a temporary Tracetest created on Kubernetes. 

### CLI

- **Unit tests**: Run by executing `make test` on `./cli` folder

### Server

- **Unit tests**: Run by executing `make test` on `./server` folder
- **Integration tests**: Run along with some unit tests running `make test` on `./server` folder, but also done in a matrix test on Github actions, by executing the `test-examples` action.
- **End-to-end tests**: Run using Tracetest to test itself. We deploy two instances of Tracetest and use one to test API calls to another on Github actions, by executing the `trace-testing` action.

**Comment by danielbdias** [comment]
Found a bug when running [Tempo Cloud example](https://github.com/kubeshop/tracetest/tree/main/examples/quick-start-grafana-cloud-tempo-nodejs). Will open a PR to solve that.

**Reply by danielbdias** [reply]
Solved on https://github.com/kubeshop/tracetest/pull/2444

**Comment by jorgeepc** [comment]
Found a new issue with data store types for `opensearch`, `elasticapm`, `signalfx ` and `newrelic`. PR to fix it here https://github.com/kubeshop/tracetest/pull/2454

**Comment by danielbdias** [comment]
Found a small configuration error on `Kubernetes and demo API` that makes PokeAPI to not report the traces correctly, I'm reopen https://github.com/kubeshop/tracetest/issues/2143 and will tackle this problem there.
